app/org/sagebionetworks/bridge/upload/IosSchemaValidationHandler.java

```diff
@@ -0,0 +1,559 @@
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import org.sagebionetworks.bridge.dao.HealthDataDao;
+import org.sagebionetworks.bridge.json.BridgeObjectMapper;
+import org.sagebionetworks.bridge.json.DateUtils;
+import org.sagebionetworks.bridge.models.healthdata.HealthDataRecordBuilder;
+import org.sagebionetworks.bridge.models.studies.Study;
+import org.sagebionetworks.bridge.models.upload.Upload;
+import org.sagebionetworks.bridge.models.upload.UploadFieldDefinition;
+import org.sagebionetworks.bridge.models.upload.UploadFieldType;
+import org.sagebionetworks.bridge.models.upload.UploadSchema;
+import org.sagebionetworks.bridge.services.UploadSchemaService;
+
+// TODO: Currently, all apps are iOS-based. However, when we start having non-iOS apps, we'll need to restructure this
+// so that it only runs in the iOS context.
+@Component
+public class IosSchemaValidationHandler implements UploadValidationHandler {
+    private static final Logger logger = LoggerFactory.getLogger(IosSchemaValidationHandler.class);
+
+    private static final Set<UploadFieldType> ATTACHMENT_TYPE_SET = EnumSet.of(UploadFieldType.ATTACHMENT_BLOB,
+            UploadFieldType.ATTACHMENT_CSV, UploadFieldType.ATTACHMENT_JSON_BLOB,
+            UploadFieldType.ATTACHMENT_JSON_TABLE);
+
+    // Note: some surveys have both questionType and questionTypeName. Some surveys only have questionType as a string.
+    // To maximize compatibility, we only check for questionType.
+    private static final Set<String> SURVEY_KEY_SET = ImmutableSet.of("endDate", "item", "questionType", "startDate");
+
+    private HealthDataDao healthDataDao;
+    private UploadSchemaService uploadSchemaService;
+
+    @Autowired
+    public void setHealthDataDao(HealthDataDao healthDataDao) {
+        this.healthDataDao = healthDataDao;
+    }
+
+    @Autowired
+    public void setUploadSchemaService(UploadSchemaService uploadSchemaService) {
+        this.uploadSchemaService = uploadSchemaService;
+    }
+
+    // iOS data comes from a third party, and we have no control over the data format. So our data validation needs to
+    // be as flexible as possible. Which means our error handling strategy is to write a warning to the logs, and then
+    // attempt to recover. This will, however, cause cascading errors further down the validation chain.
+    @Override
+    public void handle(@Nonnull UploadValidationContext context)
+            throws UploadValidationException {
+        Map<String, JsonNode> jsonDataMap = context.getJsonDataMap();
+        Map<String, byte[]> unzippedDataMap = context.getUnzippedDataMap();
+        Upload upload = context.getUpload();
+        String uploadId = upload.getUploadId();
+        Study study = context.getStudy();
+        String studyId = study.getIdentifier();
```

Comment:
I need to pass in study in a few places and studyId in a few other places. I figured it would save on typing and be more readable in the long run. Plus, the Java compile will optimize away any inefficiency from having a second local variable.